test/mechanics/constraintsPde/ConstraintCompanionPde.cpp

```diff
@@ -0,0 +1,195 @@
+{
+bool operator==(const Term& a, const Term& b)
+{
+    if (&(a.GetNode()) != &(b.GetNode()))
+        return false;
+    if (a.GetCoefficient() != b.GetCoefficient())
+        return false;
+    if (a.GetComponent() != b.GetComponent())
+        return false;
+    return true;
+}
+
+bool operator==(const Equation& a, const Equation& b)
+{
+    if (a.GetRhs(0.0) != b.GetRhs(0.0))
+        return false;
+    if (a.GetRhs(42.0) != b.GetRhs(42.0))
+        return false;
+    if (a.GetTerms() != b.GetTerms())
+        return false;
+    return true;
+}
+
+} // Constraint
+} // NuTo
+
+
+// quite the stupid tests, but oh well...
+BOOST_AUTO_TEST_CASE(PredefinedFunctions)
+{
+    auto ramp = ConstraintPde::RhsRamp(42.0, 6174.0);
+    BOOST_CHECK_EQUAL(ramp(0.0), 0.0);
+    BOOST_CHECK_EQUAL(ramp(42.0), 6174.0);
+
+    auto constant = ConstraintPde::RhsConstant(42.0);
+    BOOST_CHECK_EQUAL(constant(0.0), 42.0);
+    BOOST_CHECK_EQUAL(constant(42.0), 42.0);
+}
+
+
+// mock node, nodegroup, direction vector and lamdba function
+class Helpers
+{
+public:
+    Helpers()
+    {
+        nodes.Add(node1);
+        nodes.Add(node2);
+    }
+
+    std::function<double(double)> rhsValueFunc(double value)
+    {
+        return [value](double) { return value; };
+    }
+    std::function<double(double)> rhsValueAdjustedFunc(std::function<double(double)> f, double value)
+    {
+        return [f, value](double time) { return f(time) * value; };
+    }
+
+    std::function<double(double)> sinFunction = [](double time) { return std::sin(time); };
+    Groups::Group<NodeSimple> nodes;
+    NodeSimple node1 = NodeSimple(Eigen::VectorXd::Ones(3));
+    NodeSimple node2 = NodeSimple(Eigen::VectorXd::Ones(3));
+    double reciprocalNorm = 1.0 / std::sqrt(3.0);
+};
+
+
+BOOST_FIXTURE_TEST_CASE(Component_test, Helpers)
+{
+    // with constant RHS
+    auto eqs = ConstraintPde::Component(node1, {eDirection::X}, 1.0);
+    auto expectedEquation = ConstraintPde::Equation(node1, 0, rhsValueFunc(1.));
+    BOOST_CHECK(eqs[0] == expectedEquation);
+
+    // with lambda RHS
+    eqs = ConstraintPde::Component(node1, {eDirection::X}, sinFunction);
+    expectedEquation = ConstraintPde::Equation(node1, 0, sinFunction);
+    BOOST_CHECK(eqs[0] == expectedEquation);
+
+    // with Group
+    std::vector<ConstraintPde::Equation> expectedEquations;
+    eqs = ConstraintPde::Component(nodes, {eDirection::X}, 1.);
+    for (auto& node : nodes)
+    {
+        expectedEquations.emplace_back(node, 0, rhsValueFunc(1.));
+    }
+    BOOST_CHECK(eqs == expectedEquations);
+
+    // with Group and lambda
+    expectedEquations.clear();
+    eqs = ConstraintPde::Component(nodes, {eDirection::X}, sinFunction);
+    for (auto& node : nodes)
+    {
+        expectedEquations.emplace_back(node, 0, sinFunction);
+    }
+    BOOST_CHECK(eqs == expectedEquations);
+}
+
+
+BOOST_FIXTURE_TEST_CASE(Direction_test, Helpers)
+{
+    Eigen::Vector3d someDirection;
+    someDirection << 1, 1, 1;
```

Comment:
I find this direction test hard to follow. Can you change the direction to something more unique, say `(42, -6174, 0`)?. Or give them names. Then I could follow the single terms in the `expectedEquations` below.

A major part of the `ConstraintCompanion::Direction()` is about finding the first nonzero entry. And not constraining zero entries. These are obviously edge cases you thought about and, thus, they should be tested.

Reply:
As I said above, goal was to get the original test to run again. This is absolutely right, but at the moment I hear so much about "stupid" (have a look above) tests, that I am not sure what is considered worth of being tested and what not :laughing:

Reply:
test everything! :)
Tests are a nice opportunity to make the behavior of the code clear. Imagine a test like
~~~
BOOST_CHECK(Direction(node, Eigen::Vector3d(1,2,3), RhsConstant(5)) == Direction(node, Eigen::Vector3d(3,6,9), RhsConstant(5));
~~~
That would perfectly document the expected behavior.

Reply:
Will add this soon. At the moment I just added the edge cases with one or more zeros in the direction vector.